### ISSUE TITLE
Race condition in tests

### ISSUE THREAD

**Issue by niner** [bduggan/p6-log-async#32]
Every so often the tests fail with:
```
A react block:
  in block <unit> at t/04-filter.t line 55
Died because of the exception:
    take without gather
      in block  at t/04-filter.t line 57
```

The reason for this is the unsafe combination of gather and react in the tests. The failure can be easily reproduced by adding a sleep before the `$.source.emit` in method log. It can also easily be shown in this reduced example:

```
rakudo -e 'my $s = Supplier.new; start { sleep 1; $s.emit: 1 }; gather { react { whenever $s { take $_ } } }'
A react block:
  in block <unit> at -e line 1

Died because of the exception:
    take without gather
```

The reason for why it fails can be gathered (pun intended) from looking at the Backtrace in the whenever block:
```
nine@sphinx:~> rakudo -e 'my $s = Supplier.new; start { sleep 1; $s.emit: 1 }; gather { react { whenever $s { say Backtrace.new.full; take $_ } } }'
  in method new at SETTING::src/core.c/Backtrace.pm6 line 93
  in block  at -e line 1
  in block  at SETTING::src/core.c/Rakudo/Supply.pm6 line 446
  in method run-supply-code at SETTING::src/core.c/Rakudo/Supply.pm6 line 445
  in block  at SETTING::src/core.c/Rakudo/Supply.pm6 line 408
  in block  at SETTING::src/core.c/Supply-factories.pm6 line 153
  in block  at SETTING::src/core.c/Supply-factories.pm6 line 117
  in method run-under-recursion-list at SETTING::src/core.c/Lock/Async.pm6 line 204
  in method run-with-updated-recursion-list at SETTING::src/core.c/Lock/Async.pm6 line 183
  in method protect-or-queue-on-recursion at SETTING::src/core.c/Lock/Async.pm6 line 146
  in block  at SETTING::src/core.c/Supply-factories.pm6 line 117
  in method emit at SETTING::src/core.c/Supplier.pm6 line 60
  in method emit at SETTING::src/core.c/Supplier.pm6 line 110
  in block  at -e line 1
  in block  at SETTING::src/core.c/Promise.pm6 line 295
  in block  at SETTING::src/core.c/ThreadPoolScheduler.pm6 line 877
  in block  at SETTING::src/core.c/ThreadPoolScheduler.pm6 line 251
  in block  at SETTING::src/core.c/ThreadPoolScheduler.pm6 line 245
  in method run-one at SETTING::src/core.c/ThreadPoolScheduler.pm6 line 242
  in block  at SETTING::src/core.c/ThreadPoolScheduler.pm6 line 284
```

As this shows, the whenever block is not executed from the thread that runs the gather/react, but from the thread that does the emit. On this thread, there is no handler for the control exception thrown by `take`, which causes it to error out.

The only reason that the test mostly works is that most of the time all emits happen before we even enter the react block. Since the events are already there, we run the whenever block from within the react block's initialization code and thus on the thread that has the control exception handler in place.

In summary, it would be safer to replace usage of gather/take here with a plain old `@results` array into which you push.

**Comment by bduggan** [comment]
Thanks for the deep-dive.  I was able to reproduce this exactly as you described, and added the fix that you recommended.
